File: Grafos/Java/Grafo.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;

public class Grafo {
    ArrayList<String> vertices;
    int matriz[][];
    int qtdVertices;
    int qtdArestas;

    public Grafo(ArrayList<String> vertices){
        this.vertices = new ArrayList<>();
        this.vertices.addAll(vertices);
        this.qtdVertices = vertices.size();
        this.matriz = new int[qtdVertices][qtdVertices];
        this.qtdArestas = 0;

        for(int i = 0; i < this.qtdVertices; i++){
            for(int j = 0; j < this.qtdVertices; j++){
                this.matriz[i][j] = 0;
            }
        }
    }

    public void montarConexoesSimetricas(ArrayList<String> linhas){
        String[] campos;
        String nomeOrigem, nomeDestino;
        int indiceOrigem, indiceDestino;
        int custo;
        for (String linha : linhas) {
            campos = linha.split("@");
            nomeOrigem = campos[0];
            nomeDestino = campos[1];
            custo = Integer.parseInt(campos[2]);

            indiceOrigem = this.vertices.indexOf(nomeOrigem);
            indiceDestino = this.vertices.indexOf(nomeDestino);

            if ((this.matriz[indiceOrigem][indiceDestino] == 0) && (this.matriz[indiceDestino][indiceOrigem] == 0)){
                this.matriz[indiceOrigem][indiceDestino] = custo;
                this.matriz[indiceDestino][indiceOrigem] = custo;
                this.qtdArestas += 2;
            }

         
        
        }
    }

    public static void lerArquivo_montarGrafo(String nomeArquivo, ArrayList<String> vertices, ArrayList<String> linhas){

        try {
            FileReader arqr = new FileReader(nomeArquivo + ".txt");
            BufferedReader leitor = new BufferedReader(arqr);
            String linha;
            String[] campos;

            while((linha = leitor.readLine()) != null){
                linha = linha.toUpperCase();
                campos = linha.split("@");
                linhas.add(linha);
                

                if(!vertices.contains(campos[0])){
                    vertices.add(campos[0]);
                }

                if(!vertices.contains(campos[1])){
                    vertices.add(campos[1]);
                }
                
            }

            vertices.sort(null);

            leitor.close();
            arqr.close();
            
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void mostrarMatriz(){
        for(int i = 0; i < this.qtdVertices; i ++){
            System.out.print("\t" + i + ":");
        }
        
        System.out.println();

        for(int i = 0; i < this.qtdVertices; i++){
            System.out.print(i + ":\t");
            for(int j = 0; j < this.qtdVertices; j++){
                System.out.print(this.matriz[i][j] + "\t");
            }
            System.out.println();
        }

        for(int i = 0; i < this.qtdVertices; i++){
            System.out.println(i + ": " + vertices.get(i));
        }
    }

    public void mostrarGrafo(){
    
        
        for(int i = 0; i < this.qtdVertices; i++){
            System.out.print(this.vertices.get(i) + ":\t");
            for(int j = 0; j < this.qtdVertices; j++){
                if(this.matriz[i][j] != 0){
                    System.out.print(this.vertices.get(j) + "\t");
                }
            }
            System.out.println("");
        }

        
    }

    public boolean temCaminhoProfundidade(String origem, String destino){
        int indiceOrigem = this.vertices.indexOf(origem);
        int indiceDestino = this.vertices.indexOf(destino);

        ArrayList<String> visitados = new ArrayList<>();
        visitados.add(origem);

        return testaCaminhoProfundidade(indiceOrigem, indiceDestino, visitados);
    }

    private boolean testaCaminhoProfundidade(int no, int destino, ArrayList<String> visitados) {
        for(int col = 0; col < this.qtdVertices; col++){
            if(this.matriz[no][col] != 0 & !visitados.contains(this.vertices.get(col))) {
                if(col == destino) return true;
                System.out.println(this.vertices.get(col));
                visitados.add(this.vertices.get(col));
                return testaCaminhoProfundidade(col, destino, visitados);
            }
        }
        return false;
    }
}

File: Grafos/AtividadeGrafo/Grafo.cs
using System;
using System.IO;
using System.Collections.Generic;

class Grafo {
    public int[,] MatrizAdjacencia;
    public List<string> vertices;
    public int qtdVertices;
    public int qtdArestas;

    public Grafo(List<string> vertices) {
        this.vertices = new List<string>(vertices);
        this.qtdVertices = vertices.Count;
        this.qtdArestas = 0;
        this.MatrizAdjacencia = new int[this.qtdVertices, this.qtdVertices];

        for (int i = 0; i < this.qtdVertices; i++) {
            for (int j = 0; j < this.qtdVertices; j++) {
                this.MatrizAdjacencia[i, j] = 0;
            }
        }
    }

    public static void LeArquivoMontaGrafo(string nomeArquivo, List<string> vertices, List<string> linhas){
        try {
            using (StreamReader arqr = new StreamReader(nomeArquivo)) {
                string linha;
                while((linha = arqr.ReadLine()) != null) {
                    // Verifica se a linha tem o formato esperado
                    string[] campos = linha.Split('@');
                    if (campos.Length != 2) {
                        Console.WriteLine($"Erro: Formato inválido na linha '{linha}'. Pulando esta linha.");
                        continue; // Pula para a próxima linha
                    }

                    string cidadeOrigem = campos[0].ToUpper();
                    string cidadeDestino = campos[1].ToUpper();

                    // Adiciona as cidades à lista de vértices, se ainda não estiverem presentes
                    if (!vertices.Contains(cidadeOrigem)) {
                        vertices.Add(cidadeOrigem);
                    }
                    if (!vertices.Contains(cidadeDestino)) {
                        vertices.Add(cidadeDestino);
                    }

                    linhas.Add(linha); 
                }

                vertices.Sort(); 
            }
        } catch (IOException e) {
            Console.WriteLine(e.Message);
        }
    }




    public void InserirArestaSimetrica(string origem, string destino) {        
        int indiceOrigem = this.vertices.IndexOf(origem);
        int indiceDestino = this.vertices.IndexOf(destino);

        if (this.MatrizAdjacencia[indiceOrigem, indiceDestino] == 0) {
            this.MatrizAdjacencia[indiceOrigem, indiceDestino] = 1;
            this.qtdArestas++;
        } if (this.MatrizAdjacencia[indiceDestino, indiceOrigem] == 0) {
            this.MatrizAdjacencia[indiceDestino, indiceOrigem] = 1;
            this.qtdArestas++;
        }
    }

    public void MostrarGrafo(){
        for(int i = 0; i < this.qtdVertices; i++){
            Console.Write("(" + this.vertices[i] + ") tem conexao(oes) com" + ": \t\t");
            for(int j = 0; j < this.qtdVertices; j++){
                if(this.MatrizAdjacencia[i, j] != 0){
                    Console.Write(this.vertices[j] + " \t");
                }
            }
            Console.WriteLine("");
        }
    }

    public int MostrarGrau(string cidade) {
        int indice = this.vertices.IndexOf(cidade);
        if (indice == -1) return -27;
        int qtd = 0;
        for (int i = 0; i < this.qtdVertices; i++) {
            if (this.MatrizAdjacencia[indice, i] != 0) {
                qtd++;
            }
            if (this.MatrizAdjacencia[i, indice] != 0) {
                qtd++;
            }
        }
        return qtd;
    }

    public void MontarConexoesSimetricas(List<string> linhas){
        foreach (string linha in linhas) {
            string[] campos = linha.Split('@');
            if (campos.Length != 2) {
                Console.WriteLine($"Erro: Formato inválido na linha '{linha}'. Pulando esta linha.");
                continue; // Pula para a próxima linha
            }

            string nomeOrigem = campos[0].Trim().ToUpper();
            string nomeDestino = campos[1].Trim().ToUpper();

            int indiceOrigem = this.vertices.IndexOf(nomeOrigem);
            int indiceDestino = this.vertices.IndexOf(nomeDestino);

            // Verifica se encontrou os índices válidos para os vértices
            if (indiceOrigem != -1 && indiceDestino != -1) {
                // Adiciona conexão de indiceOrigem para indiceDestino
                if (this.MatrizAdjacencia[indiceOrigem, indiceDestino] == 0) {
                    this.MatrizAdjacencia[indiceOrigem, indiceDestino] = 1;
                    this.qtdArestas++;
                }

                // Adiciona conexão de indiceDestino para indiceOrigem
                if (this.MatrizAdjacencia[indiceDestino, indiceOrigem] == 0) {
                    this.MatrizAdjacencia[indiceDestino, indiceOrigem] = 1;
                    this.qtdArestas++;
                }
            } else {
                Console.WriteLine($"Erro: Vértices não encontrados para a linha '{linha}'. Pulando esta linha.");
            }
        }
    }


}
